internal/auth/interface.go

```diff
@@ -0,0 +1,197 @@
+//
+// Copyright 2023 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"sync"
+
+	"github.com/lestrrat-go/jwx/v2/jwt"
+)
+
+// Identity represents a particular user's identity in a particular trust domain
+// (represented by an IdentityProvider).
+type Identity struct {
+	// UserID is a stable unique identifier for the user.  This may be a large
+	// integer or a UUID, rather than something human-readable.
+	//
+	// For KeyCloak, this is `sub`.
+	UserID string
+	// HumanName is a human-readable name.  Because humans are fickle, these may
+	// not be unique or stable over time, though they should be unique at any
+	// particular time.  For example, Alex may change their handle from
+	// "alexsmith" to "alexawesome" after a life change, and someone else might
+	// enroll the "alexsmith" handle.  If you are storing data, you want UserID,
+	// not HumanName.  If you are presenting data, you probably want HumanName.
+	//
+	// For KeyCloak, this is `preferred_username`.
+	HumanName string
+	// Provider is the identity provider that vended this identity.  Note that
+	// UserID and HumanName are only unique within the context of a single
+	// identity provider.
+	Provider IdentityProvider
```

Comment:
Do we want to add email to this? For some IdPs we wouldn't quite have a preferred user name and the email would be more relevant.

Reply:
I think email _is_ the preferred username for some providers.  For others (e.g. GitHub Actions) users may not have email addresses.

Reply:
Makes sense, I think that's fine